### PR TITLE
Conform conversion of non-binary input in skeletonize functions

### LINKED ISSUE #7090: skeletonize_3d output pixel values are doubled from input pixel values
### Description:

`skeletonize` seems to preserve the pixel value of the input image. This is fine.

`skeletonize_3d` seems to *double* the pixel value of the input image. Not sure if this is considered a bug, but depending on the pixel value and pixel data type it can overflow and be confusing.

### Way to reproduce:

Slightly modified from [example](https://scikit-image.org/docs/stable/auto_examples/edges/plot_skeleton.html):
```python
from skimage.morphology import skeletonize, skeletonize_3d
from skimage import data
import matplotlib.pyplot as plt
from skimage.util import invert

image = invert(data.horse()) * 55

skeleton = skeletonize(image)
skeleton_3d = skeletonize_3d(image)

print(skeleton.max())
print(skeleton_3d.max())
```

Output:
```
55
110
```

### Version information:

```Shell
3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)]
Windows-10-10.0.19045-SP0
scikit-image version: 0.21.0
numpy version: 1.24.3
```
```

### PR DESCRIPTION
**Pull request by lagru**
## Description

Our skeletonize API is currently dealing in multiple ways with non-binary input images. Sometimes it's just silently turned into a binary array and sometimes `img_as_ubyte` is used (which scales). 

<strike>This PR adds a RuntimeWarning everywhere if the input image isn't binary, and conforms the conversion to `image.astype(bool, copy=False)`. This is probably the least controversial solution to #7090.</strike>

This PR conforms the behavior. All functions convert non-binary input implicitly with `.astype(bool, copy=False)`. 

What do you think about deprecating the implicit conversion and raising a TypeError instead? That would make testing and maintenance a lot easier! 

Closes #7090.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
Remove unexpected value scaling in `skimage.morphology.skeletonize_3d` for non-binary input images. This `skeletonize_3d` now always returns a binary array like similar functions.
```

### CODE REVIEW

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
It's not as uncontroversial as you think @lagru 😂 as I generally like the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) ("be conservative in what you send, be liberal in what you accept"), and warnings are annoying, but I'm certainly happy to listen to the criticisms section there, so happy to go with this solution.

I don't think we need to deprecate in 1.0, but maybe we can just remove it in skimage2?

**Comment by lagru** [comment]
*Theoretical rambling ahead*

I sympathize with the robustness principle from a short-term user perspective. But from a maintainer perspective and long-term user perspective, I don't think "be liberal in what you accept" is a good trade-off in a lot of cases. It makes the API surface larger which can be problematic for all kinds of reasons, not the least because it makes API changes more difficult. Documentation of parameters and behavior is generally more complex and the number of (unforseen) edge cases increases as well.

---

That said, if I think about this issue again I mainly hate the inconsistency between behavior, and documentation. Either we accept non-binary arrays and document that or we don't. Users will start to rely on either behavior anyway. So I changed my opinion and don't think the warning is the proper way to deal with this.

**Comment by lagru** [comment]
I propose accepting non-binary images for now as implied by the docstring of `skeletonize` already but conforming the behavior of all those functions to `.astype(bool, copy=False)`. But 

I'll the "convert non-binary images implicitly" issue as an item to the [list of API changes for skimage2](https://github.com/scikit-image/scikit-image/wiki/API-changes-for-skimage2) so that we can discuss it again and in context of the full library.

**Review by mkcor — COMMENTED** [review]
Thanks for addressing this inconsistency, @lagru!

**Comment by jni** [comment]
> But I'll the "convert non-binary images implicitly" issue as an item to the [list of API changes for skimage2](https://github.com/scikit-image/scikit-image/wiki/API-changes-for-skimage2) so that we can discuss it again and in context of the full library.

That's great from my perspective, @lagru. It's amusing that shortly after I came here defending the robustness principle, [it bit us in the ass in napari](https://forum.image.sc/t/struggle-to-add-property-of-unequal-length-to-labels-layer/86201/2). That's karma for you! 😂 So I now 100% agree with strongly typing this in skimage2. But it's a good one to postpone/deprecate slowly rather than to just warn.

File: skimage/morphology/_skeletonize.py [hunk 1]
```diff
@@ -253,7 +255,8 @@ def thin(image, max_num_iter=None):
     Parameters
     ----------
     image : binary (M, N) ndarray
-        The image to be thinned.
+        The image to be thinned. Zeros represent background, nonzero values are
+        foreground.
```
[the remark this answers is not in the document]
Comment (lagru) [reply skimage/morphology/_skeletonize.py]: The "by considering that" sounded a bit strange to me so I modified your suggestion a bit. Let me know if you are happy with https://github.com/scikit-image/scikit-image/pull/7095/commits/311db9dc4d1f7f3daca394df3ad977ddb677b394.

[the remark this answers is not in the document]
Comment (mkcor) [reply skimage/morphology/_skeletonize.py]: I totally admit that "by considering that" wasn't tip-top! Let me see https://github.com/scikit-image/scikit-image/commit/311db9dc4d1f7f3daca394df3ad977ddb677b394... I quite like it, I'll suggest minor adjustments.

File: skimage/morphology/_skeletonize.py [hunk 2]
```diff
@@ -852,7 +364,10 @@ def medial_axis(image, mask=None, return_distance=False, *, rng=None):
     Parameters
     ----------
     image : binary ndarray, shape (M, N)
-        The image of the shape to be skeletonized.
+        The image of the shape to skeletonize. If this input isn't already a
+        binary image it is converted into one, in which case zero values are
+        considered as the background (False) and any nonzero value as the
+        foreground (True).
```
[the remark this answers is not in the document]
Comment (lagru) [reply skimage/morphology/_skeletonize.py]: Done. Not sure what precisely you are referring to as in "juxtaposition" though.

[the remark this answers is not in the document]
Comment (lagru) [reply skimage/morphology/_skeletonize.py]: Ah, thanks for clarifying!

**Comment by mkcor** [comment]
@lagru in the context of these functions, are thin/thinned and skeletonize/skeletonized strict synonyms?

**Comment by lagru** [comment]
I was asking that myself. I am not sure. From Wikipedia on [Topological skeleton](https://en.wikipedia.org/wiki/Topological_skeleton)

> In the technical literature, the concepts of skeleton and [medial axis](https://en.wikipedia.org/wiki/Medial_axis) are used interchangeably by some authors,[[1]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-1)[[2]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-gonzales-2) while some other authors[[3]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-jain-3)[[4]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-4)[[5]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-sethian-5) regard them as related, but not the same. Similarly, the concepts of skeletonization and [thinning](https://en.wikipedia.org/wiki/Thinning_(morphology)) are also regarded as identical by some,[[2]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-gonzales-2) and not by others.[[3]](https://en.wikipedia.org/wiki/Topological_skeleton#cite_note-jain-3)

So, I suggest to use "thin" at least in `thin` and `skeletonize` in `medial_axis`. That keeps the former status quo.

**Review by jni — APPROVED** [review]

**Review by mkcor — APPROVED** [review]

**Pull request merged**